xdsl/passes.py

```diff
@@ -1,20 +1,123 @@
 from abc import ABC, abstractmethod
+from dataclasses import dataclass, Field
+from types import UnionType, NoneType
+
+from xdsl.utils.hints import isa
 from xdsl.dialects import builtin
 from xdsl.ir import MLContext
-from typing import ClassVar
+from typing import ClassVar, TypeVar, get_origin, Any, Union, get_args
+from xdsl.utils.parse_pipeline import (
+    PipelinePassSpec,
+    PassArgListType,
+    PassArgElementType,
+)
+
+_T = TypeVar("_T", bound="ModulePass")
 
 
+@dataclass
 class ModulePass(ABC):
     """
     A Pass is a named rewrite pass over an IR module.
 
     All passes are expected to leave the IR in a valid state after application.
     That is, the IR verifies. In turn, all passes can expect the IR they are
-    applied to to be in a valid state.
+    applied to be in a valid state.
```

Comment:
```suggestion
    applied to to be in a valid state.
```
:)

Reply:
ooooooh geez. My IDE suggested that change... I think we still may want to reword that.

`In turn, all passes can expect that the IR they are applied to is in a valid state.`

Reply:
I like your version more. It might be clearer to reword the whole paragraph.

"The IR must be in a valid state before and after each pass is applied, meaning that a call to `verify` must succeed after a pass has executed, but may not succeed during the execution of the pass."

Reply:
I rewrote that part, I think we should mention `.verify()` in the section. Feel free to iterate on that version!